vowpalwabbit/interactions.cc

```diff
@@ -4,123 +4,165 @@
-      }
-      if (interactions.active_interactions.find({*jt, *jt}) == interactions.active_interactions.end())
-      {
-        interactions.interactions.push_back({*jt, *jt});
-        interactions.active_interactions.insert({*jt, *jt});
-      }
-      if (interactions.leave_duplicate_interactions &&
-          interactions.active_interactions.find({*jt, *it}) == interactions.active_interactions.end())
-      {
-        interactions.interactions.push_back({*jt, *it});
-        interactions.active_interactions.insert({*jt, *it});
+          // recurrent value calculations
+          for (size_t i = 0; i < fs.size(); ++i)
+          {
+            const float x = fs.values[i] * fs.values[i];
+
+            if (!PROCESS_SELF_INTERACTIONS(fs.values[i]))
+            {
+              for (size_t j = order_of_inter - 1; j > 0; --j) results[j] += results[j - 1] * x;
+
+              results[0] += x;
+            }
+            else
+            {
+              results[0] += x;
+
+              for (size_t j = 1; j < order_of_inter; ++j) results[j] += results[j - 1] * x;
+
+              ++cnt_ft_value_non_1;
+            }
+          }
+
+          sum_feat_sq_in_inter *= results[order_of_inter - 1];  // will be explained in http://bit.ly/1Hk9JX1
+
+          // let's calculate  the number of a new features
+
+          // if number of features is less than  order of interaction then go to the next interaction
+          // as you can't make simple combination of interaction 'aaa' if a contains < 3 features.
+          // unless one of them has value != 1. and we are counting them.
+          const size_t ft_size = fs.size();
+          if (cnt_ft_value_non_1 == 0 && ft_size < order_of_inter)
+          {
+            num_features_in_inter = 0;
+            break;
+          }
+
+          size_t n;
+          if (cnt_ft_value_non_1 == 0)  // number of generated simple combinations is C(n,k)
+          {
+            n = static_cast<size_t>(
+                VW::math::choose(static_cast<int64_t>(ft_size), static_cast<int64_t>(order_of_inter)));
+          }
+          else
+          {
+            n = 0;
+            for (size_t l = 0; l <= order_of_inter; ++l)
+            {
+              // C(l+m-1, l) * C(n-m, k-l)
+              size_t num = (l == 0) ? 1 : static_cast<size_t>(VW::math::choose(l + cnt_ft_value_non_1 - 1, l));
+
+              if (ft_size - cnt_ft_value_non_1 >= order_of_inter - l)
+                num *= static_cast<size_t>(VW::math::choose(ft_size - cnt_ft_value_non_1, order_of_inter - l));
+              else
+                num = 0;
+
+              n += num;
+            }
+
+          }  // details on http://bit.ly/1Hk9JX1
+
+          num_features_in_inter *= n;
+
+          ns += order_of_inter - 1;  // jump over whole block
+        }
       }
+
+      if (num_features_in_inter == 0) continue;  // signal that values should be ignored (as default value is 1)
```

Comment:
nit: if line 126 is the only way to reach this state, same as above. Reset both variables and just do the increment?